File: deprecated/src/main/java/it/niedermann/nextcloud/deck/deprecated/database/dao/AccessControlDao.java
package it.niedermann.nextcloud.deck.database.dao;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Query;

import java.util.List;
import java.util.Set;

import it.niedermann.nextcloud.deck.model.AccessControl;

@Dao
public interface AccessControlDao extends GenericDao<AccessControl> {

    @Query("SELECT * FROM AccessControl WHERE accountId = :accountId and id = :remoteId")
    LiveData<AccessControl> getAccessControlByRemoteId(final long accountId, final long remoteId);

    @Query("SELECT * FROM AccessControl WHERE accountId = :accountId and boardId = :localBoardId and status <> 3")
    LiveData<List<AccessControl>> getAccessControlByLocalBoardId(final long accountId, final long localBoardId);

    @Query("SELECT * FROM AccessControl WHERE accountId = :accountId and boardId = :localBoardId and status <> 3")
    List<AccessControl> getAccessControlByLocalBoardIdDirectly(final long accountId, final long localBoardId);

    @Query("SELECT * FROM AccessControl WHERE accountId = :accountId and id = :remoteId")
    AccessControl getAccessControlByRemoteIdDirectly(final long accountId, final long remoteId);

    @Query("SELECT * FROM AccessControl WHERE accountId = :accountId and boardId = :boardId and (status<>1 or id is null or lastModified <> lastModifiedLocal)")
    List<AccessControl> getLocallyChangedAccessControl(long accountId, long boardId);

    @Query("SELECT distinct boardId FROM AccessControl WHERE accountId = :accountId and (status<>1 or id is null or lastModified <> lastModifiedLocal)")
    List<Long> getBoardIDsOfLocallyChangedAccessControl(long accountId);

    @Query("DELETE FROM AccessControl WHERE boardId = :localBoardId and localId not in (:idsToKeep)")
    void deleteAccessControlsForBoardWhereLocalIdsNotInDirectly(long localBoardId, Set<Long> idsToKeep);
}

File: data/local/src/main/java/it/niedermann/nextcloud/deck/data/local/dao/AccessControlDao.kt
package it.niedermann.nextcloud.deck.data.local.dao

import androidx.room3.Dao
import androidx.room3.Insert
import androidx.room3.OnConflictStrategy
import androidx.room3.Query
import io.reactivex.rxjava3.core.Flowable
import it.niedermann.nextcloud.deck.data.local.entity.AccessControlEntity
import java.util.concurrent.CompletableFuture

@Dao
interface AccessControlDao : GenericDao<AccessControlEntity> {

    @Query("SELECT * FROM AccessControl WHERE boardId = :boardId")
    fun getAclByBoard(boardId: Long): Flowable<List<AccessControlEntity>>

    @Query("SELECT * FROM AccessControl WHERE accountId = :accountId AND status != 1")
    fun getChangedAcl(accountId: Long): CompletableFuture<List<AccessControlEntity>>

    @Query("SELECT * FROM AccessControl WHERE accountId = :accountId AND remoteId = :remoteId")
    fun getAclByRemoteId(accountId: Long, remoteId: Long): CompletableFuture<AccessControlEntity?>

    @Query("SELECT * FROM AccessControl WHERE boardId = :boardId AND userId = :userId AND status != 3")
    fun getAclByBoardAndUser(boardId: Long, userId: Long): CompletableFuture<AccessControlEntity?>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insertOrReplace(entity: AccessControlEntity): CompletableFuture<Long>

    @Query("DELETE FROM AccessControl WHERE boardId = :boardId")
    fun deleteByBoardId(boardId: Long): CompletableFuture<Void?>

    @Query("DELETE FROM AccessControl WHERE localId = :localId")
    fun deleteById(localId: Long): CompletableFuture<Void?>
}
